### PR TITLE
#0: Ensure libc++ used across both projects

### PR DESCRIPTION
**Pull request by blozano-tt**
### Ticket
NA

### Problem description
Fix linker problem
UMD does not ingest `stdlib` interface library from `tt-metal` anymore.
Need to get the two projects to use same version of C++ standard library.
Solution is to adjust `CXX_FLAGS`.

### What's changed
Remove `stdlib` interface library, use `CXX_FLAGS`.
Also removed `DFMT_HEADER_ONLY` define since it comes for free, when you link the right target.
Had to fix a couple tangential issues that popped up.

### Extra info
Compiler warning (werror) -> error popped up for the below.
I don't know why it popped up, but this is illegal in C++17, so I removed constexpr.
```
old:        constexpr Segment(std::span<word_t const> contents, address_t addr, offset_t bss) :
new:         inline Segment(std::span<word_t const> contents, address_t addr, offset_t bss) :
```

### Checklist
- [x] Post commit CI passes : https://github.com/tenstorrent/tt-metal/actions/runs/11610163942
- [x] Blackhole Post commit (if applicable) : Not needed, just a build system change
- [x] Model regression CI testing passes (if applicable): Not needed, just a build system change
- [x] Device performance regression CI testing passes (if applicable) : Not needed, just a build system change
- [x] New/Existing tests provide coverage for changes : Not needed

### CODE REVIEW

**Comment by broskoTT** [comment]
Note: We will bump an UMD commit which is not on main branch, due to other changes already present on main. The #13949 which will follow right after will get it back on main branch.

File: tt_metal/llrt/tt_elffile.hpp [hunk 1]
```diff
@@ -31,7 +31,7 @@ class ElfFile {
         offset_t bss = 0;                  // words of BSS
 
        public:
-        constexpr Segment(std::span<word_t const> contents, address_t addr, offset_t bss) :
+         inline Segment(std::span<word_t const> contents, address_t addr, offset_t bss) :
```
Comment (tt-rkim) [inline tt_metal/llrt/tt_elffile.hpp:34]: formatting?

**Review by tt-rkim — APPROVED** [review]

**Comment by tt-rkim** [comment]
Is this idiomatic?

To override `CMAKE_CXX_FLAGS` to inject it everywhere? @afuller-TT

**Comment by afuller-TT** [comment]
> Is this idiomatic?
> 
> To override `CMAKE_CXX_FLAGS` to inject it everywhere? @afuller-TT

No, I don't think so.
Generally a toolchain file would be doing this type of thing, but setting the `*_INIT` variant.  These vars become cache variables (so devs may override, but also changes aren't picked up without clearing them from the local cache first).  For how we're treating things, we want `add_compiler_options` which will get plugged in without setting a cache variable.
CC: @blozano-tt

File: CMakeLists.txt [hunk 2]
```diff
@@ -285,14 +268,7 @@ target_compile_options(compiler_flags INTERFACE -DARCH_${ARCH_NAME_DEF})
 
 add_library(metal_header_directories INTERFACE)
 target_include_directories(metal_header_directories INTERFACE ${PROJECT_SOURCE_DIR}/tt_metal/hw/inc)
-target_include_directories(
-    metal_header_directories
-    SYSTEM
-    INTERFACE
-        ${reflect_SOURCE_DIR}
-        ${flatbuffers_include_dir}
-        ${nng_include_dir}
```
Comment (afuller-TT) [inline CMakeLists.txt:294]: Great to see these two deleted.  Were they already dead?  I don't see them added as deps elsewhere in this PR.

File: CMakeLists.txt [hunk 3]
```diff
@@ -39,6 +39,33 @@ project(
...
+    endif()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -stdlib=libc++")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -stdlib=libc++ -lc++abi")
+    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -stdlib=libc++ -lc++abi")
+else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsized-deallocation")
+endif()
+
+set(CMAKE_CXX_FLAGS_RELEASE "-O3")
+set(CMAKE_CXX_FLAGS_DEBUG "-O0 -g -DDEBUG")
```
Comment (afuller-TT) [inline CMakeLists.txt:61]: Do we know whether all the code tests this define only for existence and not for value?

File: CMakeLists.txt [hunk 4]
```diff
@@ -39,6 +39,33 @@ project(
...
+    find_library(LIBC++ABI c++abi)
+    if(NOT LIBC++ OR NOT LIBC++ABI)
+        message(
+            FATAL_ERROR
+            "libc++ or libc++abi not found. Make sure you have libc++ and libc++abi installed and in your PATH"
+        )
+    endif()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -stdlib=libc++")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -stdlib=libc++ -lc++abi")
+    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -stdlib=libc++ -lc++abi")
```
Comment (afuller-TT) [inline CMakeLists.txt:55]: Pretty sure we'd be better served with `add_compiler_options` and `add_link_options` to set it at the directory scope and not worry about cache vars (which I think already bit you while refactoring this).

File: CMakeLists.txt (hunk 3 above)
Comment (blozano-tt) [reply CMakeLists.txt:61]:
```
find . -name "*.h" -exec grep -H "DEBUG" {} \; | grep "#if"
find . -name "*.hpp" -exec grep -H "DEBUG" {} \; | grep "#if"
find . -name "*.cpp" -exec grep -H "DEBUG" {} \; | grep "#if"
```
Did not see any cases where the value was checked, but the output is quite noisy.

File: CMakeLists.txt (hunk 2 above)
Comment (blozano-tt) [reply CMakeLists.txt:294]: These were the two vars that were backpropagated from UMD repo into parent scope.

New UMD repo does not try to do that.

Include paths can propagate with the target, if necessary.

File: tt_metal/llrt/tt_elffile.hpp (hunk 1 above)
Comment (blozano-tt) [reply tt_metal/llrt/tt_elffile.hpp:34]: went from 1 space indentation (which I've never seen), to 2 space.

we can do a c++ formatting pass later.

File: CMakeLists.txt (hunk 4 above)
Comment (blozano-tt) [inline CMakeLists.txt:55]:
```suggestion
    add_compile_options(-stdlib=libc++)
    add_link_options(-stdlib=libc++ -lc++abi)
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -39,6 +39,32 @@ project(
...
+    if(NOT LIBC++ OR NOT LIBC++ABI)
+        message(
+            FATAL_ERROR
+            "libc++ or libc++abi not found. Make sure you have libc++ and libc++abi installed and in your PATH"
+        )
+    endif()
+    add_compile_options(-stdlib=libc++)
+    add_link_options(-stdlib=libc++ -lc++abi)
+else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsized-deallocation")
```
Comment (afuller-TT) [inline CMakeLists.txt:56]:
```suggestion
    add_compile_options(-fsized-deallocation)
```

File: CMakeLists.txt (hunk 2 above)
Comment (afuller-TT) [reply CMakeLists.txt:294]: Ouch.  So this code existed in tt-metal purely for umd 🙃 
Thanks!

File: CMakeLists.txt [hunk 6]
```diff
@@ -95,7 +113,6 @@ set(UMD_HOME "${PROJECT_SOURCE_DIR}/tt_metal/third_party/umd")
 option(WITH_PYTHON_BINDINGS "Enables build of python bindings" ON)
 option(ENABLE_CODE_TIMERS "Enable code timers" OFF)
 option(ENABLE_TRACY "Enable Tracy Profiling" OFF)
-option(ENABLE_LIBCXX "Enable using libc++" ON)
```
Comment (afuller-TT) [inline CMakeLists.txt:98]: I checked locally.  The refinements during code review weren't enough to allow it to build without libc++.  So, yeah, that config was probably dead a while back.  Removing the option is more honest/transparent.

**Review by afuller-TT — APPROVED** [review]

File: tt_metal/third_party/umd
Comment (broskoTT) [inline tt_metal/third_party/umd:1]: Don't check in without https://github.com/tenstorrent/tt-umd/commit/357154e078258810da1e84d74556cb4d0c0cde64

Comment (broskoTT) [reply tt_metal/third_party/umd:1]: I've added this commit to a branch https://github.com/tenstorrent/tt-umd/commits/brosko/tt_metal_cmake_fix_l1_membar/
This is specific branch used only temporarily for tt_metal. Will switch back to main branch in #13949

**Review by abhullar-tt — APPROVED** [review]

**Review by SeanNijjar — APPROVED** [review]

**Pull request merged**